compiler/hash-typecheck/src/new/passes/symbol_resolution.rs

```diff
@@ -120,8 +276,644 @@ impl<'tc> SymbolResolutionPass<'tc> {
+                    name,
+                    name_span,
+                    self.get_current_context_kind(),
+                )?;
+                Ok(self.context().get_binding(binding_symbol).unwrap())
+            }
+        }
+    }
+
+    /// Make [`ArgsId`] from an AST argument group, with holes for all the
+    /// arguments.
+    fn make_args_from_ast_arg_group(&self, group: &AstArgGroup) -> ArgsId {
+        macro_rules! make_args_from_ast_args {
+            ($args:expr) => {
+                self.param_ops().create_args($args.iter().enumerate().map(|(i, arg)| {
+                    return ArgData {
+                        target: arg
+                            .name
+                            .as_ref()
+                            .map(|name| ParamTarget::Name(name.ident))
+                            .unwrap_or_else(|| ParamTarget::Position(i)),
+                        value: self.new_term_hole(),
+                    };
+                }))
+            };
+        }
+        match group {
+            AstArgGroup::ExplicitArgs(args) => make_args_from_ast_args!(args),
+            AstArgGroup::ImplicitArgs(args) => make_args_from_ast_args!(args),
+        }
+    }
+
+    /// Make [`DefArgsId`] from a list of AST argument groups, using
+    /// `make_args_from_ast_arg_group` to make each argument group.
+    fn make_def_args_from_ast_arg_groups(
+        &self,
+        groups: &[AstArgGroup],
+        originating_params: DefParamsId,
+    ) -> DefArgsId {
+        self.param_ops().create_def_args(groups.iter().enumerate().map(|(index, group)| {
+            DefArgGroupData {
+                args: self.make_args_from_ast_arg_group(group),
+                param_group: (originating_params, index),
+            }
+        }))
+    }
+
+    /// Wrap a term in a function call, given a list of arguments as a list of
+    /// [`AstArgGroup`].
+    ///
+    /// Creates a new [`TermId`], which is a term that represents the
+    /// function call, and might be nested depending on `args`.
+    fn wrap_term_in_fn_call_from_ast_args(&self, subject: TermId, args: &[AstArgGroup]) -> TermId {
+        let mut current_subject = subject;
+        for arg_group in args {
+            let args = self.make_args_from_ast_arg_group(arg_group);
+            current_subject = self.new_term(Term::FnCall(FnCallTerm {
+                subject: current_subject,
+                args,
+                implicit: matches!(arg_group, AstArgGroup::ImplicitArgs(_)),
+            }));
+        }
+        current_subject
+    }
+
+    /// Apply the given list of AST arguments to the given definition
+    /// parameters, checking that the lengths match in the process.
+    ///
+    /// If successful, returns the [`DefArgsId`] that represents the arguments.
+    ///
+    /// Otherwise, returns an error.
+    fn apply_ast_args_to_def_params(
+        &self,
+        def_params: DefParamsId,
+        args: &[AstArgGroup],
+    ) -> TcResult<DefArgsId> {
```

Comment:
In the future, it could be interesting to experiment with using an error id rather than returning the error, this error points to the actual error and can be read instead of propagating it up the call stack. This has the advantage of making the return size only 8 or 12 bytes instead of however big `TcError` is. This change wouldn't be overly difficult to do, and if in the unlikely scenario that the caller wishes to read the error, they still can.

Reply:
I agree with this, maybe something to look into later on.